sys/riscv/pmap.c

```diff
@@ -1,168 +1,708 @@
+  pmap_zero_page(pg);
+  return pg;
 }
 
-void pmap_delete(pmap_t *pmap) {
-  panic("Not implemented!");
+void pmap_zero_page(vm_page_t *pg) {
+  bzero((void *)pg_dmap_addr(pg), PAGESIZE);
 }
 
-void pmap_bootstrap(paddr_t pd_pa, vaddr_t pd_va) {
-  dmap_paddr_base = kenv_get_ulong("mem_start");
-  dmap_paddr_end = kenv_get_ulong("mem_end");
-  size_t dmap_size = dmap_paddr_end - dmap_paddr_base;
-  kernel_pde = pd_pa;
+void pmap_copy_page(vm_page_t *src, vm_page_t *dst) {
+  memcpy((void *)pg_dmap_addr(dst), (void *)pg_dmap_addr(src), PAGESIZE);
+}
 
-  /* Assume physical memory starts at the beginning of L0 region. */
-  assert(is_aligned(dmap_paddr_base, L0_SIZE));
+/*
+ * Address space identifiers management.
+ */
+
+static asid_t alloc_asid(void) {
+  int free = 0;
+  WITH_SPIN_LOCK (&asid_lock) {
+    bit_ffc(asid_used, MAX_ASID, &free);
+    if (free < 0)
+      panic("Out of asids!");
+    bit_set(asid_used, free);
+  }
+  klog("alloc_asid() = %d", free);
+  return free;
+}
 
-  /* We must have enough virtual addresses. */
-  assert(dmap_size <= DMAP_MAX_SIZE);
+static void free_asid(asid_t asid) {
+  klog("free_asid(%d)", asid);
+  SCOPED_SPIN_LOCK(&asid_lock);
+  bit_clear(asid_used, (unsigned)asid);
+  tlb_invalidate_asid(asid);
+}
 
-  /* For RV32 we assume 32-bit physical address space. */
-  assert(dmap_paddr_base < dmap_paddr_end);
+/*
+ * Physical-to-virtual entries are managed for all pageable mappings.
+ */
+
+static void pv_add(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv = pool_alloc(P_PV, M_ZERO);
+  pv->pmap = pmap;
+  pv->va = va;
+  TAILQ_INSERT_TAIL(&pg->pv_list, pv, page_link);
+  TAILQ_INSERT_TAIL(&pmap->pv_list, pv, pmap_link);
+}
 
-  klog("Physical memory range: %p - %p", dmap_paddr_base, dmap_paddr_end - 1);
+static pv_entry_t *pv_find(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv;
+  TAILQ_FOREACH (pv, &pg->pv_list, page_link) {
+    if (pv->pmap == pmap && pv->va == va)
+      return pv;
+  }
+  return NULL;
+}
 
-  klog("dmap range: %p - %p", DMAP_VADDR_BASE, DMAP_VADDR_BASE + dmap_size - 1);
+static void pv_remove(pmap_t *pmap, vaddr_t va, vm_page_t *pg) {
+  assert(mtx_owned(&pv_list_lock));
+  pv_entry_t *pv = pv_find(pmap, va, pg);
+  TAILQ_REMOVE(&pg->pv_list, pv, page_link);
+  TAILQ_REMOVE(&pmap->pv_list, pv, pmap_link);
+  pool_free(P_PV, pv);
+}
 
-  /* Build direct map using superpages. */
-  pd_entry_t *pde = (void *)pd_va;
-  size_t idx = L0_INDEX(DMAP_VADDR_BASE);
-  for (paddr_t pa = dmap_paddr_base; pa < dmap_paddr_end; pa += L0_SIZE, idx++)
-    pde[idx] = PA_TO_PTE(pa) | PTE_KERN;
+/*
+ * Routines for accessing page table entries.
+ */
 
-  __sfence_vma();
+static paddr_t pmap_alloc_pde(pmap_t *pmap, vaddr_t va) {
+  assert(mtx_owned(&pmap->mtx));
 
-  void *fdtp = (void *)phys_to_dmap(FDT_get_physaddr());
-  FDT_changeroot(fdtp);
+  vm_page_t *pg = pmap_pagealloc();
+  assert(pg);
```

Comment:
`pmap_pagealloc` already zeros page and doesn't check address so you will get memory fault earlier.